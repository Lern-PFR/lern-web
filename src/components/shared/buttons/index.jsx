import PropTypes from 'prop-types';
import {
	primary,
	standard,
	subtle,
	dropdown,
	danger,
	iconButton,
	floatingActionButton,
	primaryLink,
	outlined,
	subtleLink,
} from 'theme/buttonStyles';
import DynamicButtonComponent from './DynamicButtonComponent';
import DynamicLinkButtonComponent from './DynamicLinkButtonComponent';

// @TODO: write link and subtleLink buttons once navigation has been set up.

/**
 * @name PrimaryButton
 * @description Button component to use as Call to action. Should only be used once per page (header not included).
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const PrimaryButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...primary} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name PrimaryLinkButton
 * @description Link button component to use as Call to action. Should only be used once per page (header not included).
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const PrimaryLinkButton = ({ children, ...otherProps }) => (
	<DynamicLinkButtonComponent linkStyle={primaryLink} {...primary} {...otherProps}>
		{children}
	</DynamicLinkButtonComponent>
);

/**
 * @name OutlinedButton
 * @description Button component to use as primary action when a CTA is already defined.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const OutlinedButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent linkStyle={primaryLink} {...outlined} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name OutlinedLinkButton
 * @description Link button component to use as primary when a CTA is already defined.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children The children to display inside this button.
 */
const OutlinedLinkButton = ({ children, ...otherProps }) => (
	<DynamicLinkButtonComponent linkStyle={primaryLink} {...outlined} {...otherProps}>
		{children}
	</DynamicLinkButtonComponent>
);

/**
 * @name StandardButton
 * @description Standard button component to be used for most use-cases.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const StandardButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...standard} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name SubtleButton
 * @description Button component to be used along a primary button for less crucial actions such as "cancel".
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const SubtleButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...subtle} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name SubtleLinkButton
 * @description Link button component to be used along a primary button for less used redirections.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const SubtleLinkButton = ({ children, ...otherProps }) => (
	<DynamicLinkButtonComponent linkStyle={subtleLink} {...subtle} {...otherProps}>
		{children}
	</DynamicLinkButtonComponent>
);

/**
 * @name DropdownButton
 * @description Button component to be used for dropdown menu trigger.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const DropdownButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...dropdown} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name DangerButton
 * @description Button component to use for final confirmation to a destructive action such as deleting data.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const DangerButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...danger} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name IconButton
 * @description Button component intended to display only the icon it contains, but behave as a regular button.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const IconButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...iconButton} {...otherProps} isIconOnly>
		{children}
	</DynamicButtonComponent>
);

/**
 * @name FloatingActionButton
 * @description Button component intended to display a floating action button.
 *
 * @author Timothée Simon-Franza
 *
 * @param {*} children : The children to display inside this button.
 */
const FloatingActionButton = ({ children, ...otherProps }) => (
	<DynamicButtonComponent {...floatingActionButton} {...otherProps}>
		{children}
	</DynamicButtonComponent>
);

PrimaryButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

PrimaryLinkButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

OutlinedButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

OutlinedLinkButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

StandardButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

SubtleButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

SubtleLinkButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

DropdownButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

DangerButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

IconButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

FloatingActionButton.propTypes = {
	children: PropTypes.oneOfType([
		PropTypes.element,
		PropTypes.elementType,
		PropTypes.node,
	]).isRequired,
};

export {
	FloatingActionButton,
	OutlinedButton,
	OutlinedLinkButton,
	PrimaryButton,
	PrimaryLinkButton,
	StandardButton,
	SubtleButton,
	SubtleLinkButton,
	DropdownButton,
	DangerButton,
	IconButton,
};
