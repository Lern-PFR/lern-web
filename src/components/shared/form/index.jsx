import FieldsetComponent from './FieldsetComponent';
import LabeledCheckbox from './LabeledCheckbox';
import LabeledRadioButton from './LabeledRadioButton';
import LabelComponent from './LabelComponent';
import LabeledInput from './LabeledInput';

export {
	LabeledCheckbox,
	FieldsetComponent,
	LabeledRadioButton,
	LabelComponent,
	LabeledInput,
};
