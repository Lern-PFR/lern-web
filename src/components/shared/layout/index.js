import MainLayout from './MainLayout';

export {
	// eslint-disable-next-line import/prefer-default-export
	MainLayout,
};
