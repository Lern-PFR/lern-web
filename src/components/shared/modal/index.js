import Modal from './Modal';
import ModalFooter from './ModalFooter';
import PromptModal from './PromptModal';
import useModal from './useModal';

export {
	Modal,
	ModalFooter,
	PromptModal,
	useModal,
};
