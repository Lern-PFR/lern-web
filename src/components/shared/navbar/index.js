import Navbar from './Navbar';
import navbarLink from './NavbarLink';

export {
	Navbar as default,
	navbarLink,
};
