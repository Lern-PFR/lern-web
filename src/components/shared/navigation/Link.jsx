import PropTypes from 'prop-types';
import { Link as ReactRouterLink } from 'react-router-dom';
import styled from 'styled-components';
import { color } from 'styled-system';

const StyledLink = styled(ReactRouterLink)(
	{
		cursor: 'inherit',
		color: 'inherit',
		fontFamily: 'inherit',
		fontSize: 'inherit',
		fontWeight: 'inherit',
		textDecoration: 'inherit',
	},
	color,
);

/**
 * @name Link
 * @description Renders a link to an internal URL, by providing keys path to the route.
 *
 * @author Yann Hodiesne
 *
 * @param {string}	to			key of the route to target with this Link
 * @param {bool}	[disabled]	Whether the link is disabled or not. If so, the onClick event is aborted.
 */
const Link = ({ to, disabled, ...props }) => (
	<StyledLink to={to} {...(disabled ? { onClick: (e) => e.preventDefault() } : {})} {...props} />
);

Link.propTypes = {
	to: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.object,
		PropTypes.func,
	]),
	disabled: PropTypes.bool,
};

Link.defaultProps = {
	disabled: false,
};

Link.defaultProps = {
	to: '',
};

export default Link;
