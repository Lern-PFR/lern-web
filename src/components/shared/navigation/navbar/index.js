import Navbar from './Navbar';
import NavbarLink from './NavbarLink';
import NavbarLogo from './NavbarLogo';

export {
	Navbar as default,
	NavbarLink,
	NavbarLogo,
};
