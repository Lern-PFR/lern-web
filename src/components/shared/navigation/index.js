import Link from './Link';
import Navbar from './navbar';

export {
	Link,
	Navbar,
};
