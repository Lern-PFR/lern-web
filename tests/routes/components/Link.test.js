import { Link as ReactRouterLink, StaticRouter } from 'react-router-dom';
import { mount, shallow } from 'enzyme';
import { Link } from 'components/shared/navigation';

describe('Link', () => {
	it('should match previoys snapshot', () => {
		const sut = <Link to="invalid_key">Hello, world!</Link>;
		const wrapper = shallow(sut);

		expect(wrapper).toMatchSnapshot();
	});

	it('should render with a ReactRouter Link', () => {
		const sut = <Link to="invalid_key">Hello, world!</Link>;
		const wrapper = mount(<StaticRouter>{sut}</StaticRouter>);

		expect(wrapper.containsMatchingElement(<ReactRouterLink to="invalid_key">Hello, world!</ReactRouterLink>)).toEqual(true);
	});
});
