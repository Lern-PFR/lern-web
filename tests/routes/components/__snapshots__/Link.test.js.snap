// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should match previoys snapshot 1`] = `
<Styled(Link)
  to="invalid_key"
>
  Hello, world!
</Styled(Link)>
`;
